imap-codec/src/fragmentizer.rs

```diff
@@ -0,0 +1,1401 @@
+                            end: self.end,
+                            announcement: Some(announcement),
+                            ending: LineEnding::Lf,
+                        });
+                        LatestByte::Other
+                    }
+                    b'{' => LatestByte::OpeningBracket,
+                    _ => LatestByte::Other,
+                },
+                LatestByte::Cr { announcement } => match next_byte {
+                    b'\r' => LatestByte::Cr { announcement: None },
+                    b'\n' => {
+                        parsed_line = Some(FragmentInfo::Line {
+                            start: self.start,
+                            end: self.end,
+                            announcement,
+                            ending: LineEnding::CrLf,
+                        });
+                        LatestByte::Other
+                    }
+                    b'{' => LatestByte::OpeningBracket,
+                    _ => LatestByte::Other,
+                },
+            };
+
+            if parsed_line.is_some() {
+                // We parsed a complete line
+                break;
+            }
+        }
+
+        (parsed_byte_count, parsed_line)
+    }
+}
+
+/// The latest byte seen by the [`LineParser`] with additional accumulated state.
+#[derive(Clone, Debug)]
+enum LatestByte {
+    Other,
+    OpeningBracket,
+    Digit {
+        length: u32,
+    },
+    Plus {
+        length: u32,
+    },
+    ClosingBracket {
+        announcement: LiteralAnnouncement,
+    },
+    Cr {
+        announcement: Option<LiteralAnnouncement>,
+    },
+}
+
+/// Stateful parser for the next literal fragment.
+#[derive(Clone, Debug)]
+struct LiteralParser {
+    /// Where we started parsing the literal.
+    start: usize,
+    /// Until where we parsed the literal.
+    end: usize,
+    /// Remaining bytes we need to parse.
+    remaining: u32,
+}
+
+impl LiteralParser {
+    fn new(start: usize, length: u32) -> Self {
+        Self {
+            start,
+            end: start,
+            remaining: length,
+        }
+    }
+
+    fn parse(&mut self, unprocessed_bytes: &VecDeque<u8>) -> (usize, Option<FragmentInfo>) {
+        if unprocessed_bytes.len() < self.remaining as usize {
+            // Not enough bytes yet
+            let parsed_byte_count = unprocessed_bytes.len();
+            self.end += parsed_byte_count;
+            self.remaining -= parsed_byte_count as u32;
+            (parsed_byte_count, None)
+        } else {
+            // There are enough bytes
+            let parsed_byte_count = self.remaining as usize;
+            self.end += parsed_byte_count;
+            self.remaining = 0;
+            let parsed_literal = FragmentInfo::Literal {
+                start: self.start,
+                end: self.end,
+            };
+            (parsed_byte_count, Some(parsed_literal))
+        }
+    }
+}
+
+pub fn parse_tag(message_bytes: &[u8]) -> Option<Tag> {
```

Comment:
I made this public by accident.
```suggestion
fn parse_tag(message_bytes: &[u8]) -> Option<Tag> {
```